blockchain/blockdao/block_migrate.go

```diff
@@ -0,0 +1,393 @@
+	h := blk.HashBlock()
+	dao.batch.Put(blockNS, topHeightKey, heightValue, "failed to put top height")
+	dao.batch.Put(blockNS, topHashKey, h[:], "failed to put top hash")
+	topDB, _, err := dao.getDBFromIndex(0)
+	if err != nil {
+		return err
+	}
+	if err := topDB.WriteBatch(dao.batch); err != nil {
+		return err
+	}
+
+	// write tipHeight into chain.db
+	dao.batch.Put(blockNS, topHeightKey, heightValue, "failed to put top height")
+	dao.batch.Put(blockNS, topHashKey, h[:], "failed to put top hash")
+	if err = dao.kvstore.WriteBatch(dao.batch); err != nil {
+		return err
+	}
+
+	return nil
+	//return os.Remove(path.Dir(dao.cfg.DbPath) + "/oldchain.db")
+}
+
+func (dao *blockDAO) getBlockByHeightLegacy(height uint64) (*block.Block, error) {
+	h, err := dao.getBlockHashLegacy(height)
+	if err != nil {
+		return nil, err
+	}
+	blk, err := dao.getBlockLegacy(h)
+	if err != nil {
+		return nil, err
+	}
+	receipts, err := dao.getReceiptsLegacy(height)
+	if err != nil {
+		return nil, err
+	}
+	blk.Receipts = receipts
+	return blk, nil
+}
+
+func (dao *blockDAO) commitForMigration(kvstore db.KVStore) error {
+	if err := dao.blkStore.Commit(); err != nil {
+		return err
+	}
+	if err := dao.receiptStore.Commit(); err != nil {
+		return err
+	}
+	if err := dao.hashStoreTop.Commit(); err != nil {
+		return err
+	}
+	if err := dao.hashStore.Commit(); err != nil {
+		return err
+	}
+	return kvstore.WriteBatch(dao.batch)
+}
+
+func (dao *blockDAO) putBlockForMigration(blk *block.Block) error {
+	blkHeight := blk.Height()
+	h, err := dao.getBlockHashLegacy(blkHeight)
+	if err != nil {
+		return err
+	}
+	serBlk, err := blk.Serialize()
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize block")
+	}
+	if dao.compressBlock {
+		timer := dao.timerFactory.NewTimer("compress_header")
+		serBlk, err = compress.Compress(serBlk)
+		timer.End()
+		if err != nil {
+			return errors.Wrapf(err, "error when compressing a block")
+		}
+	}
+	if err = dao.blkStore.Add(serBlk, true); err != nil {
+		return err
+	}
+	// write receipts
+	if blk.Receipts != nil {
+		receipts := iotextypes.Receipts{}
+		for _, r := range blk.Receipts {
+			receipts.Receipts = append(receipts.Receipts, r.ConvertToReceiptPb())
+		}
+		if receiptsBytes, err := proto.Marshal(&receipts); err == nil {
+			if err = dao.receiptStore.Add(receiptsBytes, true); err != nil {
+				return err
+			}
+		} else {
+			log.L().Error("failed to serialize receipits for block", zap.Uint64("height", blkHeight))
+		}
+	}
+	h = blk.HashBlock()
+	if err = dao.hashStoreTop.Add(h[:], true); err != nil {
+		return nil
+	}
+	if err = dao.hashStore.Add(h[:], true); err != nil {
+		return nil
+	}
+
+	heightValue := byteutil.Uint64ToBytes(blkHeight)
+	hashKey := hashKey(h)
+	dao.batch.Put(blockHashHeightMappingNS, hashKey, heightValue, "failed to put hash -> height mapping")
+	return nil
+}
+
+func (dao *blockDAO) getBlockHashLegacy(height uint64) (hash.Hash256, error) {
```

Comment:
these code are all code for read legacy format DB, could be removed once migration is done